### ISSUE TITLE
[EKS] [request]: Setting the STS Session name

### ISSUE THREAD

**Issue by taer** [aws/containers-roadmap#2362]
I think this is probably more of a bug report, but I cannot find the proper channels.
Version Info:
EKS: 1.29
Eks Pod ID agent: v1.2.0-eksbuild.1

We are using eks-pod-identity. It's been working great till we started using IAM based kafka.

I directly hit the eks-pod-identity pod inside a container agent via
```
       AWS_CONTAINER_CREDENTIALS_FULL_URI:      http://169.254.170.23/v1/credentials                                                                                                        
       AWS_CONTAINER_AUTHORIZATION_TOKEN_FILE:  /var/run/secrets/pods.eks.amazonaws.com/serviceaccount/eks-pod-identity-token                                                               │
```

I took the resulting AccessKey, SecertKey, and Token, and used them to do a `aws sts get-caller-identity`

```
{
    "UserId": "AROAXYZP66II6MSBPLOUU:eks-k8s-wl-dev-engine-boo-5af5e7ac-5754-49ea-b28f-2c2a2eb95fbb",
    "Account": "BLAH",
    "Arn": "arn:aws:sts::BLAH:assumed-role/dev-use1-default-bookingApiPod-20240529171038526700000001/eks-k8s-wl-dev-engine-boo-5af5e7ac-5754-49ea-b28f-2c2a2eb95fbb"
}
```

The SessionName is non static, and I can't find any way to force set it. The issue comes from the MSK IAM usage. MSK doesn't allow "reauthentication". We get this error when the session name changes
```
failed authentication due to: Cannot change principals during re-authentication from IAM.arn:aws:sts::BLAH:assumed-role/prd-use1-default-bookingApiPod-20240529174141238800000002/eks-k8s-wl-prd-engine-boo-064f1ed1-2349-4774-b895-9a69ccc3eeb1: IAM.arn:aws:sts::BLAH:assumed-role/prd-use1-default-bookingApiPod-20240529174141238800000002/eks-k8s-wl-prd-engine-boo-875fdfcd-19cd-4ac3-8544-7077f94a6e39
```

Most services we've used IAM for to date don't care. The solution normally would be to set `AWS_ROLE_SESSION_NAME` when calling STS:AssumeRole.   But we're not calling that, the eks-pod-identity pod is. The ContainerCredentialProvider in the SDK is just calling the `AWS_CONTAINER_CREDENTIALS_FULL_URI` URL with the contents of `AWS_CONTAINER_AUTHORIZATION_TOKEN_FILE` as the Authorization header. There is no option as far as I've found to have that include a constant session name to satisfy MSK.

Thanks!

**Comment by prateekgogia** [comment]
> I directly hit the eks-pod-identity pod inside a container agent via

IIUC, you are manually calling the `AWS_CONTAINER_CREDENTIALS_FULL_URI` with the token from inside a pod running on your node and use the credentials to make this `aws sts get-caller-identity` cli call.

Trying to understand if these credentials ever work from inside a Pod when an application is trying to reach Kafka and fail when using these credentials manual with aws cli? or they never work with Kafka?

**Comment by taer** [comment]
Yeah. basically, I was doing what the SDK does manually just to validate.

The credentials given by the pod-identity work perfectly w/ Kafka until they reach their expiration.  When the SDK goes back to the pod-identity-agent near the expiration, the agent refreshes the tokens for it. The issue though is it refreshes it with a new sessionName. And the MSK IAM auth scheme hates that.    It considers the session-name change to be a change in principal, and screams
 `failed authentication due to: Cannot change principals during re-authentication from IAM.arn:aws:sts::BLAH:assumed-role/prd-use1-default-bookingApiPod-20240529174141238800000002/eks-k8s-wl-prd-engine-boo-064f1ed1-2349-4774-b895-9a69ccc3eeb1: IAM.arn:aws:sts::BLAH:assumed-role/prd-use1-default-bookingApiPod-20240529174141238800000002/eks-k8s-wl-prd-engine-boo-875fdfcd-19cd-4ac3-8544-7077f94a6e39`

important parts in there is the change here
```
arn:aws:sts......roleName/eks-k8s-wl-prd-engine-boo-064f1ed1-2349-4774-b895-9a69ccc3eeb1
````
```
arn:aws:sts......roleName/eks-k8s-wl-prd-engine-boo-875fdfcd-19cd-4ac3-8544-7077f94a6e39
````

That sessionName of pod-name-uuid is generated by the eks-pod-identity-agent. Ideally, we could configure the agent to just use the pod-name and not include the uuid, and we'd be golden.

**Comment by dims** [comment]
fyi code is here now - https://github.com/aws/eks-pod-identity-agent - can we please move this to an issue there? 🙏🏾

**Comment by taraspos** [comment]
I've opened feature request asking for support of Custom Session Tags and Custom Session Names in the above repo:
- https://github.com/aws/eks-pod-identity-agent/issues/14

**Comment by metropolis-ameer** [comment]
following updates to this issue

**Comment by Santiacmaestre** [comment]
We ran into the exact same issue with **EKS Pod Identity + MSK IAM authentication**.
Everything works fine initially, but once the **eks-pod-identity-agent** refreshes credentials, the **session name changes**, and MSK rejects the re-auth with:

```
Cannot change principals during re-authentication
```

This happens because the Pod Identity agent generates a **new STS session name** on each credential refresh, and there’s currently **no way to set a static session name** in the agent.

### Workaround (validated in Java / Spring Boot)
While waiting for an official fix, we stabilized this by introducing an **assumed IAM role** dedicated to MSK access, with a **fixed session name** defined at the application level.

### Steps:
1. **Create a dedicated IAM role** for Kafka/MSK access (produce/consume).
     - Move the MSK permissions to this role.
2. **Add a trust policy** that allows your **Pod Identity role** to `sts:AssumeRole` into the dedicated MSK role.
3. **In your app config**, explicitly set:
   * `awsRoleArn` for the new dedicated role
   * A **fixed `awsRoleSessionName`** (so the effective principal never changes).

#### Example (for Spring Boot / Spring Cloud Stream Kafka binder):

```yaml
spring:
  cloud:
    stream:
      kafka:
        binder:
          configuration:
              sasl:
                # login.callback.handler.class: software.amazon.msk.auth.iam.IAMOAuthBearerLoginCallbackHandler
                mechanism: AWS_MSK_IAM
                client.callback.handler.class: software.amazon.msk.auth.iam.IAMClientCallbackHandler
                jaas.config: software.amazon.msk.auth.iam.IAMLoginModule required awsRoleArn="arn:aws:iam::[account-id]:role/[role-name]" awsRoleSessionName="[static-session-name]"  awsStsRegion="[region-name]";
                security.protocol: SASL_SSL
```

#### Why this works
- Pod Identity continues to handle the credential rotation and token refresh, but every connection to MSK now **assumes the same dedicated IAM role** with a **constant session name**.
- From MSK’s perspective, the authenticated principal stays consistent, so re-authentication succeeds.

#### Results

- Verified with multiple Java/Spring Boot services on EKS 1.29 + Pod Identity Agent v1.2.0.
- Ran >8h soak tests — **no re-authentication errors** after the change.
- No need to migrate to IRSA; this workaround keeps Pod Identity in place.

**In short:**
Until eks-pod-identity-agent supports a static session name, assuming a dedicated MSK role with a fixed session name is a reliable workaround for Java-based workloads.

<img width="1120" height="499" alt="Image" src="https://github.com/user-attachments/assets/29d81dcb-83ab-4f5d-8873-69d3333643e5" />

**Comment by edikmkoyan** [comment]
> We ran into the exact same issue with **EKS Pod Identity + MSK IAM authentication**. Everything works fine initially, but once the **eks-pod-identity-agent** refreshes credentials, the **session name changes**, and MSK rejects the re-auth with:
> 
> ```
> Cannot change principals during re-authentication
> ```
> 
> This happens because the Pod Identity agent generates a **new STS session name** on each credential refresh, and there’s currently **no way to set a static session name** in the agent.
> 
> ### Workaround (validated in Java / Spring Boot)
> While waiting for an official fix, we stabilized this by introducing an **assumed IAM role** dedicated to MSK access, with a **fixed session name** defined at the application level.
> 
> ### Steps:
> 1. **Create a dedicated IAM role** for Kafka/MSK access (produce/consume).
>    
>    * Move the MSK permissions to this role.
> 2. **Add a trust policy** that allows your **Pod Identity role** to `sts:AssumeRole` into the dedicated MSK role.
> 3. **In your app config**, explicitly set:
>    
>    * `awsRoleArn` for the new dedicated role
>    * A **fixed `awsRoleSessionName`** (so the effective principal never changes).
> 
> #### Example (for Spring Boot / Spring Cloud Stream Kafka binder):
> spring:
>   cloud:
>     stream:
>       kafka:
>         binder:
>           configuration:
>               sasl:
>                 # login.callback.handler.class: software.amazon.msk.auth.iam.IAMOAuthBearerLoginCallbackHandler
>                 mechanism: AWS_MSK_IAM
>                 client.callback.handler.class: software.amazon.msk.auth.iam.IAMClientCallbackHandler
>                 jaas.config: software.amazon.msk.auth.iam.IAMLoginModule required awsRoleArn="arn:aws:iam::[account-id]:role/[role-name]" awsRoleSessionName="[static-session-name]"  awsStsRegion="[region-name]";
>                 security.protocol: SASL_SSL
> #### Why this works
> * Pod Identity continues to handle the credential rotation and token refresh, but every connection to MSK now **assumes the same dedicated IAM role** with a **constant session name**.
> * From MSK’s perspective, the authenticated principal stays consistent, so re-authentication succeeds.
> 
> #### Results
> * Verified with multiple Java/Spring Boot services on EKS 1.29 + Pod Identity Agent v1.2.0.
> * Ran >8h soak tests — **no re-authentication errors** after the change.
> * No need to migrate to IRSA; this workaround keeps Pod Identity in place.
> 
> **In short:** Until eks-pod-identity-agent supports a static session name, assuming a dedicated MSK role with a fixed session name is a reliable workaround for Java-based workloads.
> 
> <img alt="Image" width="1120" height="499" src="https://private-user-images.githubusercontent.com/72419315/504714285-29d81dcb-83ab-4f5d-8873-69d3333643e5.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjMwNDA3MzcsIm5iZiI6MTc2MzA0MDQzNywicGF0aCI6Ii83MjQxOTMxNS81MDQ3MTQyODUtMjlkODFkY2ItODNhYi00ZjVkLTg4NzMtNjlkMzMzMzY0M2U1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTEzVDEzMjcxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQxMDkzN2RmMmYyNjQ1ZGUyMjBlNmQzOGUxYjZjNjkwZGQ2YTMwNGQxNzM3Njc3MTg4MmU5YjQzYzExNTNmY2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.uSuWGwIdI45oAVK916JMd7y4znS-SEPBSc6z16gMWqc">

This will work with a single self-assumed role as well.

**Comment by Santiacmaestre** [comment]
We tried using a single assume-role (pod identity one), but it's not working. Pod identity normally assumes the role, but it cannot be the same as the assume-role. During the refresh token process, it will use a different random session name.
